edb/edgeql/compiler/stmt.py

```diff
@@ -132,7 +132,9 @@ def compile_ForQuery(
 
             iterator_view = stmtctx.declare_view(
                 iterator, qlstmt.iterator_alias,
-                path_id_namespace=path_id_ns, ctx=scopectx)
+                new_namespace=False,
+                path_id_namespace=path_id_ns,
+                ctx=iterator_ctx or scopectx)
```

Comment:
In fact, the actual bug seems to be that `scopectx` is superfluous and only gets in the way as `declare_view` puts `path_namespace` on it only for it to get thrown away.  `declare_view` iself provides enough isolation, there's no need to surround it with fences.  Try simply unnesting the whole `with ... as scopectx` block and replace `scopectx` with `sctx`.  This also obviates the need for `bodyctx` nesting below.  The `new_namespace` hack isn't needed too.

Reply:
Great catch.

I did that, and threw some slight refactoring/rearrangement in because dropping such a big chunk of code into the same nesting level without any visual breaks was giving me the willies :P